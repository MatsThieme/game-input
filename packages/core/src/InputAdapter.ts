import type { InputAxis } from "./InputAxis/InputAxis";
import type { InputButton } from "./InputButton/InputButton";

/**
 * InputAdapters should always cache their InputButtons and InputAxes between updates
 */
export interface InputAdapter<
    ActionMappedToButton extends string | number = string | number,
    ActionMappedToAxis extends string | number = string | number,
    AdapterInputButton extends InputButton = InputButton,
    AdapterInputAxis extends InputAxis = InputAxis,
> {
    /**
     * Get an InputButton object for the specified input of the corresponding adapter.
     *
     * @param key a unique identifier for an input.
     */
    getButton(key: ActionMappedToButton): Readonly<AdapterInputButton> | undefined;

    /**
     * Get an InputAxis object for the specified input of the corresponding adapter.
     *
     * @param key a unique identifier for an input.
     */
    getAxis(key: ActionMappedToAxis): Readonly<AdapterInputAxis> | undefined;

    /**
     * @internal
     * Called once per frame.
     */
    update(): void;

    /**
     * @internal
     * Prepare the InputAdapter for disposal by removing listeners or clearing references to InputButtons/-Axis.
     */
    dispose(): void;
}
