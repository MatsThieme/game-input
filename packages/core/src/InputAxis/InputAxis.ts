type NumberTupleMinLength = [number, ...number[]];

export class InputAxis {
    private _values: NumberTupleMinLength;
    private _length?: number;
    private _changed: number;

    public constructor(values?: number | NumberTupleMinLength) {
        if (values === undefined) this._values = [0];
        else if (typeof values === "number") this._values = [values];
        else this._values = values.slice() as NumberTupleMinLength;

        this._changed = 0;
    }

    /**
     * Returns whether values changed in the last frame.
     */
    public get changed(): boolean {
        return this._changed === 1;
    }

    public getValues(): Readonly<NumberTupleMinLength> {
        return this._values;
    }

    /**
     * @internal
     */
    public setValues(values: Readonly<NumberTupleMinLength>): void {
        if (
            this._values.length !== values.length ||
            this._values.some((value, index) => value !== values[index])
        ) {
            this._values = values.slice() as NumberTupleMinLength;

            this._length !== undefined && (this._length = undefined);
            this._changed = 2;
        }
    }

    /**
     * @internal
     */
    public update(): void {
        this._changed > 0 && this._changed--;
    }

    public getLength(): number {
        if (typeof this._length === "number") {
            return this._length;
        }

        if (this._values.length === 1) {
            return (this._length = this._values[0]);
        }

        return (this._length = Math.sqrt(
            this._values.reduce((result, current) => result + current ** 2, 0)
        ));
    }
}
