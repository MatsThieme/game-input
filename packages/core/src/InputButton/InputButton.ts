export class InputButton {
    private _isSetDown: boolean | undefined;
    private _down: boolean;
    private _wasDown: boolean;

    /**
     * Store state information of a button.
     */
    public constructor() {
        this._isSetDown = false;
        this._down = false;
        this._wasDown = false;
    }

    /**
     * Returns whether the button was down in the last frame.
     */
    public get down(): boolean {
        return this._down;
    }

    /**
     * Returns whether the button was down in the second last frame.
     */
    public get wasDown(): boolean {
        return this._wasDown;
    }

    /**
     * Returns whether the button was down in the last frame and was not down in the second last frame.
     */
    public get click(): boolean {
        return this._down && !this._wasDown;
    }

    /**
     * Returns whether the button was down in the last and second last frame.
     */
    public get clicked(): boolean {
        return this._down && this._wasDown;
    }

    /**
     * @internal
     */
    public setDown(down: boolean): void {
        this._isSetDown = down;
    }

    /**
     * @internal
     */
    public update(): void {
        this._wasDown = this._down;

        if (this._isSetDown !== undefined) this._down = this._isSetDown;

        this._isSetDown = undefined;
    }
}
