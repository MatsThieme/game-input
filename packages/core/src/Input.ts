import { InputAdapter } from "./InputAdapter";
import type { InputAxis } from "./InputAxis/InputAxis";
import type { InputButton } from "./InputButton/InputButton";
import type {
    GetActionMappedTo,
    GetAdapter,
    GetAdapterAxis,
    GetAdapterAxisActions,
    GetAdapterButton,
    GetAdapterButtonActions,
    GetAdaptersForAction,
    GetInputAxisForAdapter,
    GetInputButtonForAdapter,
    MappedAxes,
    MappedButtons,
} from "./types";

export const enum InputState {
    ready = "ready",
    disposed = "disposed",
}

type AdapterFactories<Adapters> = { [K in keyof Adapters]: () => Adapters[K] };

export class Input<
    ButtonMapping extends MappedButtons<Adapters>,
    AxisMapping extends MappedAxes<Adapters>,
    Adapters extends GetAdapter<ButtonMapping, AxisMapping>,
    ButtonActions extends GetActionMappedTo<Adapters, ButtonMapping>,
    AxisActions extends GetActionMappedTo<Adapters, AxisMapping>,
> {
    private readonly _mappingButtons: Readonly<ButtonMapping>;
    private readonly _mappingAxes: Readonly<AxisMapping>;
    private readonly _adapterFactories: AdapterFactories<Adapters>;
    private readonly _adapters: Adapters;

    private readonly _frameInputButtonCache: Partial<Record<string, Readonly<InputButton>>>;
    private readonly _frameInputAxisCache: Partial<Record<string, InputAxis>>;

    private _state: InputState = InputState.ready;

    public constructor(
        adapters: AdapterFactories<Adapters>,
        mappingButtons: ButtonMapping,
        mappingAxes: AxisMapping,
    ) {
        this._mappingButtons = mappingButtons;
        this._mappingAxes = mappingAxes;
        this._adapterFactories = adapters;
        this._adapters = this._createAdapters();

        this._frameInputButtonCache = {};
        this._frameInputAxisCache = {};
    }

    /**
     * Returns an InputButton object mapped to the given input action.
     */
    public getButton<
        T extends ButtonActions,
        U extends GetInputButtonForAdapter<GetAdaptersForAction<T, ButtonMapping, Adapters>>,
    >(action: T): Readonly<U> | undefined {
        if (!this._checkState()) {
            return;
        }

        if (this._frameInputButtonCache[action]) {
            return this._frameInputButtonCache[action] as U;
        }

        let button: Readonly<U> | undefined;

        for (const adapter in this._adapters) {
            const adapterMapping = this._mappingButtons[adapter];

            if (!adapterMapping) {
                continue;
            }

            const mappedTo = adapterMapping[action];

            if (!mappedTo) {
                continue;
            }

            const buttonAdapter = this._adapters[adapter].getButton(mappedTo) as U | undefined;

            if (!buttonAdapter) {
                continue;
            }

            if (buttonAdapter.click) {
                return buttonAdapter;
            } else if (buttonAdapter.down) {
                button = buttonAdapter;
            } else if (!button) {
                button = buttonAdapter;
            }
        }

        if (!button) {
            return;
        }

        return (this._frameInputButtonCache[action] = button);
    }

    public getAdapterButton<
        AdapterName extends keyof Adapters & string,
        ActionName extends GetAdapterButtonActions<Adapters[AdapterName]>,
        AdapterButton extends GetAdapterButton<Adapters[AdapterName]>,
    >(adapter: AdapterName, action: ActionName): AdapterButton | undefined {
        if (!this._checkState()) {
            return;
        }

        const cacheKey = adapter + "." + action;

        if (this._frameInputButtonCache[cacheKey]) {
            return this._frameInputButtonCache[cacheKey] as AdapterButton;
        }

        const adapterButton = this._adapters[adapter].getButton(action) as
            | AdapterButton
            | undefined;

        this._frameInputButtonCache[cacheKey] = adapterButton;

        return adapterButton;
    }

    /**
     * Returns an InputAxis object mapped to the given InputAction.
     */
    public getAxis<
        T extends AxisActions,
        U extends GetInputAxisForAdapter<GetAdaptersForAction<T, AxisMapping, Adapters>>,
    >(action: T): Readonly<U> | undefined {
        if (!this._checkState()) {
            return;
        }

        if (this._frameInputAxisCache[action]) {
            return this._frameInputAxisCache[action] as U;
        }

        let axis: Readonly<U> | undefined;

        for (const adapter in this._adapters) {
            const adapterMapping = this._mappingAxes[adapter];

            if (!adapterMapping) {
                continue;
            }

            const mappedTo = adapterMapping[action];

            if (!mappedTo) {
                continue;
            }

            const axisAdapter = this._adapters[adapter].getAxis(mappedTo) as U | undefined;

            if (!axisAdapter) {
                continue;
            }

            if (!axis || axisAdapter.getLength() > axis.getLength()) {
                axis = axisAdapter;
            }
        }

        if (!axis) {
            return;
        }

        return (this._frameInputAxisCache[action] = axis as unknown as InputAxis) as U; // TODO: fix type
    }

    public getAdapterAxis<
        AdapterName extends keyof Adapters & string,
        ActionName extends GetAdapterAxisActions<Adapters[AdapterName]>,
        AdapterAxis extends GetAdapterAxis<Adapters[AdapterName]>,
    >(adapter: AdapterName, action: ActionName): AdapterAxis | undefined {
        if (!this._checkState()) {
            return;
        }

        const cacheKey = adapter + "." + action;

        if (this._frameInputAxisCache[cacheKey]) {
            return this._frameInputAxisCache[cacheKey] as AdapterAxis;
        }

        const adapterAxis = this._adapters[adapter].getAxis(action) as AdapterAxis | undefined;

        this._frameInputAxisCache[cacheKey] = adapterAxis;

        return adapterAxis;
    }

    public update(): void {
        if (!this._checkState()) {
            return;
        }

        for (const adapter in this._adapters) {
            this._adapters[adapter].update();
        }

        this._clearCache();
    }

    private _clearCache(): void {
        for (const cached in this._frameInputButtonCache) {
            this._frameInputButtonCache[cached] = undefined;
        }

        for (const cached in this._frameInputAxisCache) {
            this._frameInputAxisCache[cached] = undefined;
        }
    }

    private _createAdapters(): Adapters {
        const adapters: Partial<Adapters> = {};

        for (const adapter in this._adapterFactories) {
            adapters[adapter] = this._adapterFactories[adapter]();
        }

        return adapters as Adapters;
    }

    private _checkState(): boolean {
        const disposed = this._state === InputState.disposed;

        if (disposed) {
            console.error(
                "Input is unusable because it was disposed with `input.dispose()`. Create a new instance.",
            );
        }

        return !disposed;
    }

    /**
     * Dispose all input adapters.
     */
    public dispose(): void {
        this._state = InputState.disposed;

        this._clearCache();

        for (const name in this._adapters) {
            this._adapters[name].dispose();
        }
    }
}
