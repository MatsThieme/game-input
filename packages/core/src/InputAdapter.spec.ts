import { Input } from "./Input";
import { describe, it, expect } from "vitest";
import { InputAdapter } from "./InputAdapter";
import { InputAxis } from "./InputAxis/InputAxis";
import { InputButton } from "./InputButton/InputButton";

describe("InputAdapter", () => {
    it("should update correctly", () => {
        const eventSource = new ExternalEventSource();
        const input = new Input(
            {
                test: (): TestAdapter => new TestAdapter(eventSource),
            },
            {
                test: {
                    TestButton: TestAdapter.buttonKey,
                },
            },
            {
                test: {
                    TestAxis: TestAdapter.axisKey,
                },
            }
        );

        const steps = [
            (): void => eventSource.dispatch("click"),
            (): void => input.update(),
            (): void => eventSource.dispatch("click"),
            (): void => input.update(),
            (): void => input.update(),
            (): void => eventSource.dispatch("release"),
            (): void => input.update(),
            (): void => eventSource.dispatch("move", [12, 23]),
            (): void => input.update(),
            (): void => eventSource.dispatch("move", [995, 124]),
            (): void => input.update(),
            (): void => eventSource.dispatch("move", [534, 63]),
            (): void => input.update(),
            (): void => input.update(),
        ];

        for (const step of steps) {
            step();

            snapInput();
        }
        input.dispose();

        function snapInput(): void {
            expect(input.getButton("TestButton")).toMatchSnapshot();
            expect(input.getAxis("TestAxis")).toMatchSnapshot();
        }
    });
});

class TestAdapter implements InputAdapter {
    public static readonly buttonKey = "TestAdapter-buttonKey";
    public static readonly axisKey = "TestAdapter-axisKey";

    private readonly _eventSource: ExternalEventSource;

    private _button?: InputButton;
    private _axis?: InputAxis;

    private readonly _clickListener = (): void => {
        this._button?.setDown(true);
    };
    private readonly _releaseListener = (): void => {
        this._button?.setDown(false);
    };

    private readonly _moveListener = (value?: number[] | undefined): void => {
        if (value) {
            this._axis?.setValues(value);
        }
    };

    public constructor(eventSource: ExternalEventSource) {
        this._eventSource = eventSource;

        this._button = new InputButton();
        this._axis = new InputAxis();
        this._eventSource.addListener("click", this._clickListener);
        this._eventSource.addListener("release", this._releaseListener);
        this._eventSource.addListener("move", this._moveListener);
    }

    public getButton(key: "TestAdapter-buttonKey"): Readonly<InputButton> | undefined {
        if (key === TestAdapter.buttonKey) {
            return this._button;
        }
    }

    public getAxis(key: "TestAdapter-axisKey"): Readonly<InputAxis<number[]>> | undefined {
        if (key === TestAdapter.axisKey) {
            return this._axis;
        }
    }

    public update(): void {
        this._button?.update();
        this._axis?.update();
    }

    public dispose(): void {
        this._button = undefined;
        this._axis = undefined;

        this._eventSource.removeListener("click", this._clickListener);
        this._eventSource.removeListener("release", this._releaseListener);
        this._eventSource.removeListener("move", this._moveListener);
    }
}

type EventType = "click" | "release" | "move";
// eslint-disable-next-line @typescript-eslint/no-explicit-any
class ExternalEventSource<T extends (...args: any[]) => unknown = (...args: any[]) => unknown> {
    private readonly _listeners: {
        [K in EventType]: T[];
    } = {
        click: [],
        release: [],
        move: [],
    };

    public addListener(event: EventType, cb: T): void {
        this._listeners[event].push(cb);
    }

    public removeListener(event: EventType, cb: T): void {
        const index = this._listeners[event].indexOf(cb);

        if (index !== -1) {
            this._listeners[event].splice(index, 1);
        }
    }

    public dispatch<E extends EventType>(
        event: E,
        value?: E extends "move" ? number[] : undefined
    ): void {
        this._listeners[event].forEach((cb) => cb(value));
    }
}
