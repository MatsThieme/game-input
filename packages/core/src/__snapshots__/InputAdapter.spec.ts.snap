// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputAdapter > should update correctly 1`] = `
InputButton {
  "_down": false,
  "_isSetDown": true,
  "_wasDown": false,
}
`;

exports[`InputAdapter > should update correctly 2`] = `
InputAxis {
  "_changed": 0,
  "_values": [
    0,
  ],
}
`;

exports[`InputAdapter > should update correctly 3`] = `
InputButton {
  "_down": true,
  "_isSetDown": undefined,
  "_wasDown": false,
}
`;

exports[`InputAdapter > should update correctly 4`] = `
InputAxis {
  "_changed": 0,
  "_values": [
    0,
  ],
}
`;

exports[`InputAdapter > should update correctly 5`] = `
InputButton {
  "_down": true,
  "_isSetDown": true,
  "_wasDown": false,
}
`;

exports[`InputAdapter > should update correctly 6`] = `
InputAxis {
  "_changed": 0,
  "_values": [
    0,
  ],
}
`;

exports[`InputAdapter > should update correctly 7`] = `
InputButton {
  "_down": true,
  "_isSetDown": undefined,
  "_wasDown": true,
}
`;

exports[`InputAdapter > should update correctly 8`] = `
InputAxis {
  "_changed": 0,
  "_values": [
    0,
  ],
}
`;

exports[`InputAdapter > should update correctly 9`] = `
InputButton {
  "_down": true,
  "_isSetDown": undefined,
  "_wasDown": true,
}
`;

exports[`InputAdapter > should update correctly 10`] = `
InputAxis {
  "_changed": 0,
  "_values": [
    0,
  ],
}
`;

exports[`InputAdapter > should update correctly 11`] = `
InputButton {
  "_down": true,
  "_isSetDown": false,
  "_wasDown": true,
}
`;

exports[`InputAdapter > should update correctly 12`] = `
InputAxis {
  "_changed": 0,
  "_values": [
    0,
  ],
}
`;

exports[`InputAdapter > should update correctly 13`] = `
InputButton {
  "_down": false,
  "_isSetDown": undefined,
  "_wasDown": true,
}
`;

exports[`InputAdapter > should update correctly 14`] = `
InputAxis {
  "_changed": 0,
  "_values": [
    0,
  ],
}
`;

exports[`InputAdapter > should update correctly 15`] = `
InputButton {
  "_down": false,
  "_isSetDown": undefined,
  "_wasDown": true,
}
`;

exports[`InputAdapter > should update correctly 16`] = `
InputAxis {
  "_changed": 2,
  "_values": [
    12,
    23,
  ],
}
`;

exports[`InputAdapter > should update correctly 17`] = `
InputButton {
  "_down": false,
  "_isSetDown": undefined,
  "_wasDown": false,
}
`;

exports[`InputAdapter > should update correctly 18`] = `
InputAxis {
  "_changed": 1,
  "_values": [
    12,
    23,
  ],
}
`;

exports[`InputAdapter > should update correctly 19`] = `
InputButton {
  "_down": false,
  "_isSetDown": undefined,
  "_wasDown": false,
}
`;

exports[`InputAdapter > should update correctly 20`] = `
InputAxis {
  "_changed": 2,
  "_values": [
    995,
    124,
  ],
}
`;

exports[`InputAdapter > should update correctly 21`] = `
InputButton {
  "_down": false,
  "_isSetDown": undefined,
  "_wasDown": false,
}
`;

exports[`InputAdapter > should update correctly 22`] = `
InputAxis {
  "_changed": 1,
  "_values": [
    995,
    124,
  ],
}
`;

exports[`InputAdapter > should update correctly 23`] = `
InputButton {
  "_down": false,
  "_isSetDown": undefined,
  "_wasDown": false,
}
`;

exports[`InputAdapter > should update correctly 24`] = `
InputAxis {
  "_changed": 2,
  "_values": [
    534,
    63,
  ],
}
`;

exports[`InputAdapter > should update correctly 25`] = `
InputButton {
  "_down": false,
  "_isSetDown": undefined,
  "_wasDown": false,
}
`;

exports[`InputAdapter > should update correctly 26`] = `
InputAxis {
  "_changed": 1,
  "_values": [
    534,
    63,
  ],
}
`;

exports[`InputAdapter > should update correctly 27`] = `
InputButton {
  "_down": false,
  "_isSetDown": undefined,
  "_wasDown": false,
}
`;

exports[`InputAdapter > should update correctly 28`] = `
InputAxis {
  "_changed": 0,
  "_values": [
    534,
    63,
  ],
}
`;
